### PR TITLE
Fix `--make-repro-path` handling of "ILLinker compatible" parameters

### LINKED ISSUE #78752: Resolve workaround for linker 1541 main assembly rooting behavior in AOT compiler
https://github.com/dotnet/runtime/blob/2ae7f5d42c817614839f3bf2e889981fdaa03631/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/UsageBasedMetadataManager.cs#L308-L314 has a workaround for rooting behavior of illink.

This behavior has been fixed in https://github.com/dotnet/linker/issues/1541.

We should remove the workaround in AOT compiler and make the two behaviors consistent.

Per @MichalStrehovsky this will require:
* Wait until the targets get propagated to the runtime repo because AFAIK the ILLink.targets we use to build tests are the one from the SDK used to build the repo. I think deleting the logic from the compiler would make us start rooting things.
* We will also need a change in the ILCompiler targets to ignore TrimmerRootAssembly with a non-default RootMode metadata. I didn’t even know such thing existed. We don’t need to be compatible with undocumented things (or we can if needed... it’s just more work and there’s huge value in keeping things simple for us and for customers).

### LINKED ISSUE #79032: ILCompiler --make-repro-path improvements
When ILCompiler creates a repro package several options are handled incorrectly and stored in the `repro.rsp` file.
## Issues
1. Command line options: `--root`, `--conditionalroot`, `--trim` can accept assembly names or paths to assemblies as values.
If the latter is used, the `repro.rsp` file will have incorrect paths specified, since in general all paths should be converted to the relative ones (relative to the repro package). Currently, this does not happen for these options and can cause AOT compilation failures.
2. Command line options: `--conditionalroot`, `--trim` if not present in the original response file, will be emitted to the `repro.rsp` without a value, which is not acceptable by the compiler.
3. Command line options: `-o` (or `--out`) and `--exportsfile` are just copied over to the `repro.rsp`. These options can point to non-existing directories for the output files and will break the compiler.

## Suggestions
1. Use the same principle as with `-r` (or `--reference`) when a path to assembly is specified for `--root`, `--conditionalroot`, `--trim` options
2. Do not output options without a value
3. Create folders inside of the repro package for the output command-line options, and adjust their values with the relative paths 
    a) The created folders in repro package should be prefixed with `_in_` if they refer to an input option and `_out` if they refer to an output option

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
We have a couple command line options that for compat with ILLinker accept either file names, or assembly names.

The zip packages created with `--make-repro-path` don't work well when such parameters are specified because `--make-repro-path` doesn't know whether to de-illinkerify them (the logic is general-purpose and shared with other tools). If this was a file name, the generated response file would be read as an assembly name (because the file no longer exists when we run repro).

The fix is to stop accepting illinkerified command line arguments and instead shim this in MSBuild. That way no transformation is needed - these should always be treated as assembly names in ILC.

Fixes #79032.
Fixes #78752.

Cc @dotnet/ilc-contrib

### CODE REVIEW

**Review by ivanpovazan — APPROVED** [review]
LGTM! Thank you!

**Review by vitek-karas — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
Sigh, the test failure is relevant and annoying:

```
    Unhandled Exception: System.IO.FileNotFoundException: Assembly name does not match filename D:\a\_work\1\s\artifacts\tests\coreclr\obj\windows.x64.Checked\Managed\Loader\classloader\StaticVirtualMethods\DiamondShape\svm_diamondshape_d\svm_diamondshape_d.dll
       at ILCompiler.CompilerTypeSystemContext.AddModule(String, String, Boolean, CompilerTypeSystemContext.ModuleData, Boolean) + 0x3d2
       at ILCompiler.CompilerTypeSystemContext.GetModuleForSimpleName(String, Boolean) + 0xa6
       at ILCompiler.Program.Run() + 0x10fc
```

Our test tree has a bad habit of naming files differently from the name of the assembly contained it in. We have safeguards to check for this and normally don't accept it, but the entrypoint assembly doesn't go through the assembly resolution logic because we open it as a file. We'd previously do that for rooting too - we don't resolve assembly with name `svm_diamondshape_d` - we just say "open the assembly in file named `svm_diamondshape_d.dll`. After this change, we say "resolve assembly `svm_diamondshape_d`" and that goes through the validation that fails if the assembly inside `svm_diamondshape_d.dll` is actually manifested as `svm_diamondshape` (without the `_d` part).

We'll probably need some new type system API or something.

**Comment by MichalStrehovsky** [comment]
> We'll probably need some new type system API or something.

I fixed that issue by fixing another bug as well (#78752). The problem disappears because the test infra stops passing spurious `--root blah_d` arguments for assemblies that should not actually be rooted (the problem was that after my first set of changes, we'd start getting `--root blah_d` parameters for assemblies that are actually `blah` (we find this out once we open the `blah_d` file). Previously this was not a problem because instead we got `--root path/to/some/blah_d.dll` that we would just open without any assumptions about the name of the assembly.

@vitek-karas can you please have a look at the last two commits?

File: src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/MemberTypesAllOnCopyAssembly.cs [hunk 1]
```diff
@@ -65,8 +65,7 @@ namespace Mono.Linker.Tests.Cases.DataFlow
 	[KeptTypeInAssembly ("base.dll", "Mono.Linker.Tests.Cases.DataFlow.Dependencies.MemberTypesAllBaseType/PrivateNestedType")]
 	[KeptMemberInAssembly ("base.dll", "Mono.Linker.Tests.Cases.DataFlow.Dependencies.MemberTypesAllBaseType/PrivateNestedType", new string[] { "PrivateMethod()" })]
 
-	// https://github.com/dotnet/runtime/issues/78752
-	[KeptMember (".ctor()", By = Tool.Trimmer)]
+	[KeptMember (".ctor()", By = Tool.Trimmer | Tool.NativeAot)]
```
Comment (vitek-karas) [inline src/tools/illink/test/Mono.Linker.Tests.Cases/DataFlow/MemberTypesAllOnCopyAssembly.cs:68]:
```suggestion
	[KeptMember (".ctor()")]
```

We don't have Kept validation for the analyzer (it doesn't make sense).

**Review by vitek-karas — APPROVED** [review]
Looks good - thanks!

**Pull request merged**